src/main/ruby/truffleruby/core/string.rb

```diff
@@ -70,6 +70,65 @@ def byteslice(index_or_range, length = undefined)
     byteslice index, length
   end
 
+  def bytesplice(*args)
+    is_range = Primitive.is_a?(args[0], Range)
+
+    if args.size == 3
+      start = Primitive.rb_to_int(args[0])
+      start += bytesize if start < 0
+
+      len = Primitive.rb_to_int(args[1])
+      str = StringValue(args[2])
+    elsif args.size == 2
+      unless is_range
+        raise(TypeError, "wrong argument type #{Primitive.class(args[0])} (expected Range)")
+      end
+
+      start, len = Primitive.range_normalized_start_length(args[0], bytesize)
+      len = 0 if len < 0
+      str = StringValue(args[1])
+    else
+      raise(ArgumentError, "wrong number of arguments (given #{args.size}, expected 2..3)")
+    end
+
+    if len < 0
+      raise(IndexError, "negative length #{len}")
+    end
+
+    if bytesize < start || start < 0
+      if is_range
+        raise(RangeError, "#{args[0]} out of range")
+      else
+        raise(IndexError, "index #{start} out of string")
+      end
+    end
+
+    len = bytesize - start if len > bytesize - start
+    finish = start + len
+
+    unless on_codepoint_boundary?(start)
+      raise(IndexError, "offset #{start} does not land on character boundary")
+    end
+    unless on_codepoint_boundary?(finish)
+      raise(IndexError, "offset #{finish} does not land on character boundary")
+    end
+
+    Primitive.check_mutable_string(self)
+    enc = Primitive.encoding_ensure_compatible_str(self, str)
+    Primitive.string_splice(self, str, start, len, enc)
+  end
+
+  private def on_codepoint_boundary?(byte_pos)
+    char_pos = Primitive.byte_index_to_character_index(self, byte_pos)
+    adjusted_byte_pos = if char_pos >= size
+                          bytesize
```

Comment:
Not sure what case this `then` branch handles. Is it a case when `Primitive.byte_index_to_character_index` returns char index out of a String bounds?

**suggestion:** So I suppose that some internal error is raised instead of silent swallowing.